### PR TITLE
Serialize AssetName to JSON as hex string

### PR DESCRIPTION
**Pull request by cblp**
Fix (second attempt) of incorrect UTF-8 decoding of non-textual (non-UTF-8) asset names.

I also changed tx-view to test JSON repr.

BEFORE:
```js
{
  "51f056b530aba5969bf996d1be51c7aa3581f6e0d7b9e0911a5c0c1d": {
    "TheDrifter13": 1,
    "GOLDtest": 5,
    "��\"3DUfw��������": 42 // UTF-8 fail, actually
  }
}
```

AFTER:
```js
{
  "51f056b530aba5969bf996d1be51c7aa3581f6e0d7b9e0911a5c0c1d": {
    "546865447269667465723133": 1,
    "474f4c4474657374": 5,
    "00112233445566778899AABBCCDDEEFF": 42
  }
}
```

This is how it already looks like in UI (Cardano Explorer):
> asset1...
> Ticker:
> Name:
> Description:
> Policy ID: 51f056b530aba5969bf996d1be51c7aa3581f6e0d7b9e0911a5c0c1d
> Asset Name: 474f4c4474657374

Future Daedalus:
> Name: GoodCoin
> Asset name: 676f6f64636f696e (ASCII: goodcoin)

### CODE REVIEW

**Comment by gitmachtl** [comment]
Just as a note: It should also be than `<policyID><assetName>` combined hexstring for the normal query utxo output too, also for the tx-out and mint parameters for CLI. like the "subject" parameter for the metadata-registry, same format for all.

**Comment by cblp** [comment]
Removed textual AssetName everywhere, hex only

**Comment by gitmachtl** [comment]
> Removed textual AssetName everywhere, hex only

do we really need the "." separator now than for policyID.assetName ?  polID is 56 chars long, rest is the assetName if present.

**Comment by gitmachtl** [comment]
> Removed textual AssetName everywhere, hex only

minted some tokens, looking good

**Comment by gitmachtl** [comment]
I noticed in the JSON output, that the address key is now showing the bech32 address of the requested address. A few versions ago this showed the hex-representation of the address. Does not make a huge difference, but is this address field actively used somewhere? Looks like too me, that this is an additional information nobody really needs?

File: cardano-api/src/Cardano/Api/Value.hs [hunk 1]
```diff
@@ -354,9 +345,12 @@ valueFromNestedRep (ValueNestedRep bundles) =
...
   toJSON (ValueNestedRep bundles) = object $ map toPair bundles
     where
-     toPair :: ValueNestedBundle -> (Text, Aeson.Value)
-     toPair (ValueNestedBundleAda q) = ("lovelace", toJSON q)
-     toPair (ValueNestedBundle pid assets) = (renderPolicyId pid, toJSON assets)
+      toPair :: ValueNestedBundle -> (Text, Aeson.Value)
+      toPair (ValueNestedBundleAda q) = ("lovelace", toJSON q)
+      toPair (ValueNestedBundle pid assets) =
+        ( renderPolicyId pid
+        , toJSON $ Map.mapKeys serialiseToRawBytesHexText assets
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/Value.hs:352]: Let's extract this logic to a  `ToJSON AssetName` instance.

File: cardano-api/src/Cardano/Api/Value.hs [hunk 2]
```diff
@@ -366,11 +360,28 @@ instance FromJSON ValueNestedRep where
...
       parsePid :: (Text, Aeson.Value) -> Parser ValueNestedBundle
       parsePid ("lovelace", q) = ValueNestedBundleAda <$> parseJSON q
-      parsePid (pid, q) =
-        case deserialiseFromRawBytesHex AsScriptHash (Text.encodeUtf8 pid) of
-          Just sHash -> ValueNestedBundle (PolicyId sHash) <$> parseJSON q
-          Nothing -> fail $ "Failure when deserialising PolicyId: "
-                         <> Text.unpack pid
+      parsePid (pid, quantityBundleJson) = do
+        sHash <-
+          note ("Failure when deserialising PolicyId: " <> Text.unpack pid) $
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/Value.hs:365]: "Failed to deserialise PolicyId: " would be a better error message.

File: cardano-api/src/Cardano/Api/Value.hs [hunk 3]
```diff
@@ -366,11 +360,28 @@ instance FromJSON ValueNestedRep where
...
+          deserialiseFromRawBytesHex AsScriptHash $ Text.encodeUtf8 pid
+        quantityBundleRaw <- parseJSON quantityBundleJson
+        quantityBundle <-
+          Map.fromList
+          <$> for (Map.assocs quantityBundleRaw) \(nameText, quantity) -> do
+                name <-
+                  note
+                    ("Failure when deserialising AssetName"
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/Value.hs:373]: Like wise for here. We should extract this to a `FromJSON AssetName` and then implement a JSON roundtrip test.

File: cardano-api/src/Cardano/Api/Value.hs [hunk 4]
```diff
@@ -366,11 +360,28 @@ instance FromJSON ValueNestedRep where
...
+                      <> Text.unpack nameText) $
+                  deserialiseFromRawBytesHex AsAssetName $
+                  Text.encodeUtf8 nameText
+                pure (name, quantity)
+        pure $ ValueNestedBundle (PolicyId sHash) quantityBundle
+
+note :: MonadFail m => String -> Maybe a -> m a
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/Value.hs:380]: `failWith` or something similar would be a better name.

File: cardano-api/src/Cardano/Api/Value.hs [hunk 5]
```diff
@@ -406,6 +417,9 @@ renderPolicyId (PolicyId scriptHash) = serialiseToRawBytesHexText scriptHash
...
 renderAssetId :: AssetId -> Text
 renderAssetId AdaAssetId = "lovelace"
-renderAssetId (AssetId polId (AssetName assetName))
-  | BS.null assetName = renderPolicyId polId
-  | otherwise         = renderPolicyId polId <> "." <> Text.decodeUtf8 assetName
+renderAssetId (AssetId polId (AssetName "")) = renderPolicyId polId
+renderAssetId (AssetId polId assetName) =
+  renderPolicyId polId <> "." <> renderAssetName assetName
+
+renderAssetName :: AssetName -> Text
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/Value.hs:424]: `renderAssetName` is overkill. We just use `serialiseToRawBytesHexText`.

File: cardano-cli/src/Cardano/CLI/Mary/ValueParser.hs [hunk 6]
```diff
@@ -111,10 +113,15 @@ decimal = do
...
-assetName =
-    toAssetName <$> some alphaNum
-  where
-    toAssetName = AssetName . Text.encodeUtf8 . Text.pack
+assetName = do
+  hexText <- some hexDigit
+  note "AssetName deserisalisation failed" $
+    deserialiseFromRawBytesHex AsAssetName $ BSC.pack hexText
+
+note :: MonadFail m => String -> Maybe a -> m a
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Mary/ValueParser.hs:121]: Why have you re-implemented the same function twice? DRY - Do not Repeat Yourself.

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]
Nice, a few comments

File: cardano-api/src/Cardano/Api/Value.hs [hunk 7]
```diff
@@ -1,6 +1,8 @@
+{-# LANGUAGE BlockArguments #-}
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/Value.hs:1]: Let's remove this pragma. I don't think its work it for the sake of not specifying parentheses in your lamba function.

**Comment by gitmachtl** [comment]
@cblp any updates here from your side? would be nice to see this in 1.31.0 😄

File: cardano-api/src/Cardano/Api/Value.hs (hunk 7 above)
Comment (cblp) [reply cardano-api/src/Cardano/Api/Value.hs]: done

File: cardano-api/src/Cardano/Api/Value.hs (hunk 1 above)
Comment (cblp) [reply cardano-api/src/Cardano/Api/Value.hs]: done

File: cardano-api/src/Cardano/Api/Value.hs (hunk 4 above)
Comment (cblp) [reply cardano-api/src/Cardano/Api/Value.hs:369]: this name is used in Cardano.Prelude (but with a slightly different type) and https://hackage.haskell.org/package/errors-2.3.0/docs/Control-Error-Util.html#v:note. I consider it somewhat traditional.

File: cardano-cli/src/Cardano/CLI/Mary/ValueParser.hs (hunk 6 above)
Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Mary/ValueParser.hs:121]: each one has type incompatible with others

Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Mary/ValueParser.hs:121]: Parser has no MonadFail

File: cardano-api/src/Cardano/Api/Value.hs (hunk 2 above)
Comment (cblp) [reply cardano-api/src/Cardano/Api/Value.hs]: done

File: cardano-api/src/Cardano/Api/Value.hs (hunk 3 above)
Comment (cblp) [reply cardano-api/src/Cardano/Api/Value.hs]: done

File: cardano-api/src/Cardano/Api/Value.hs (hunk 5 above)
Comment (cblp) [reply cardano-api/src/Cardano/Api/Value.hs]: done

**Comment by gitmachtl** [comment]
how does the output of "Friendly output multi-asset with ASCII version where possible" look like? i would leave it up to the software behind the CLI to decide to show it as readable or as hex/binary. makes it more complicated to post-process if the output is not hex all the time.

**Comment by cblp** [comment]
> how does the output of "Friendly output multi-asset with ASCII version where possible" look like?

Here is an example:
```yaml
mint:
  policy 52dc3d43b6d2465e96109ce75ab61abe5e9c1d8a3c9ce6ff8a3af528:
    asset 736b79 (sky): 142
```

> i would leave it up to the software behind the CLI to decide to show it as readable or as hex/binary. makes it more complicated to post-process if the output is not hex all the time.

`cardano-cli transaction view` is an end-user software that uses Cardano API.

API version of cardano-cli is to be built in #3213

**Comment by gitmachtl** [comment]
ah, so the normal `utxo query` is not affected by this, only in the `transaction view` output?

**Comment by cblp** [comment]
@gitmachtl `utxo query` is affected only in part of switching from text to hex.

**Comment by gitmachtl** [comment]
> @gitmachtl `utxo query` is affected only in part of switching from text to hex.

thats perfect, so hex output in the normal `utxo query` output, also in the JSON version of the output. and friendly names in the `transaction view` output.

**Comment by cblp** [comment]
@dcoutts should I fix docs (couttscoins etc.) in this PR?

File: cardano-cli/src/Cardano/CLI/Mary/ValueParser.hs (hunk 6 above)
Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Mary/ValueParser.hs:121]: Sorry, Aeson.Parser has MonadFail, but not MonadError, which is used by Cardano.Prelude.note.

And I can't figure out the right place for MonadFail version, so I repeated the code to save ourselves from bad module structure.

Can you suggest a place for this function?

**Comment by gitmachtl** [comment]
ping 😄

File: cardano-cli/src/Cardano/CLI/Mary/ValueParser.hs (hunk 6 above)
Comment (Jimbo4350) [reply cardano-cli/src/Cardano/CLI/Mary/ValueParser.hs:121]: I don't follow. The `note` in `cardano-api/src/Cardano/Api/Value.hs` has the type signature `note :: MonadFail m => String -> Maybe a -> m a` which is identical to the `note` in `cardano-cli/src/Cardano/CLI/Mary/ValueParser.hs`.

Why not put `note` in `Cardano.Api.Utils`?

File: cardano-api/test/Test/Cardano/Api/Typed/Value.hs [hunk 8]
```diff
@@ -65,7 +69,21 @@ canonicalise =
...
+  property $ do
+    v <- forAll genAssetName
+    tripping (Map.singleton v ()) encode eitherDecode
+
+
 -- -----------------------------------------------------------------------------
 
 tests :: TestTree
-tests = fromGroup $$discover
+tests = $testGroupGenerator
```
Comment (Jimbo4350) [inline cardano-api/test/Test/Cardano/Api/Typed/Value.hs:89]: Why the change to `testGroupGenerator`?

**Review by Jimbo4350 — CHANGES_REQUESTED** [review]
A few comments

File: cardano-api/src/Cardano/Api/SerialiseUsing.hs [hunk 9]
```diff
@@ -63,33 +63,40 @@ instance SerialiseAsRawBytes a => Show (UsingRawBytesHex a) where
...
+    where
+      tname  = (tyConName . typeRepTyCon . typeRep) (Proxy :: Proxy a)
+
+instance SerialiseAsRawBytes a => ToJSONKey (UsingRawBytesHex a) where
+  toJSONKey =
+    Aeson.toJSONKeyText $ \(UsingRawBytesHex x) -> serialiseToRawBytesHexText x
+
+instance
+  (SerialiseAsRawBytes a, Typeable a) => FromJSONKey (UsingRawBytesHex a)
+  where
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/SerialiseUsing.hs:84]: Strange indent here.

`(SerialiseAsRawBytes a, Typeable a) => FromJSONKey (UsingRawBytesHex a) where` should be fine

**Comment by Jimbo4350** [comment]
> @dcoutts should I fix docs (couttscoins etc.) in this PR?
Yes, please update the docs with respect to the serialization change.

File: cardano-api/src/Cardano/Api/SerialiseUsing.hs [hunk 10]
```diff
@@ -63,33 +63,40 @@ instance SerialiseAsRawBytes a => Show (UsingRawBytesHex a) where
...
-        case Base16.decode (Text.encodeUtf8 str) of
-          Right raw -> case deserialiseFromRawBytes ttoken raw of
-            Just x  -> return (UsingRawBytesHex x)
-            Nothing -> fail ("cannot deserialise " ++ show str)
-          Left msg  -> fail ("invalid hex " ++ show str ++ ", " ++ msg)
-      where
-        ttoken = proxyToAsType (Proxy :: Proxy a)
-        tname  = (tyConName . typeRepTyCon . typeRep) (Proxy :: Proxy a)
-
-instance SerialiseAsRawBytes a => ToJSONKey (UsingRawBytesHex a)
```
[the remark this answers is not in the document]
Comment (cblp) [reply cardano-api/src/Cardano/Api/SerialiseUsing.hs:91]: The default ToJSONKey implementation is the Value-based encoding; we need the Text-based here.

File: cardano-api/test/Test/Cardano/Api/Typed/Value.hs (hunk 8 above)
Comment (cblp) [reply cardano-api/test/Test/Cardano/Api/Typed/Value.hs:89]: There is already grouping ability in `tasty` library, we don't need our home-grown version.

File: cardano-cli/src/Cardano/CLI/Mary/ValueParser.hs (hunk 6 above)
Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Mary/ValueParser.hs:121]: `Cardano.Api.Utils` is internal in `cardano-api`, I use `Cardano.CLI.Helpers`.

File: cardano-api/src/Cardano/Api/Value.hs [hunk 11]
```diff
@@ -70,13 +72,14 @@ import           Data.String (IsString (..))
 import           Data.Text (Text)
 import qualified Data.Text as Text
 import qualified Data.Text.Encoding as Text
+-- import           Data.Traversable (for)
```
[the remark this answers is not in the document]
Comment (cblp) [reply cardano-api/src/Cardano/Api/Value.hs]: removed

File: cardano-api/test/Test/Cardano/Api/Typed/Value.hs (hunk 8 above)
Comment (cblp) [reply cardano-api/test/Test/Cardano/Api/Typed/Value.hs]: Indeed, this change is lateral, reverted. It should be done more generally in a separate PR.

File: cardano-cli/src/Cardano/CLI/Mary/ValueParser.hs (hunk 6 above)
Comment (cblp) [reply cardano-cli/src/Cardano/CLI/Mary/ValueParser.hs]: Code moved to `Cardano.Api.Utils`

File: cardano-api/src/Cardano/Api/SerialiseUsing.hs (hunk 9 above)
Comment (Jimbo4350) [reply cardano-api/src/Cardano/Api/SerialiseUsing.hs]: The limit is 100 chars as per the style document

File: cardano-api/src/Cardano/Api/SerialiseUsing.hs [hunk 12]
```diff
@@ -63,33 +63,39 @@ instance SerialiseAsRawBytes a => Show (UsingRawBytesHex a) where
...
+    Aeson.toJSONKeyText $ \(UsingRawBytesHex x) -> serialiseToRawBytesHexText x
+
+instance
+  (SerialiseAsRawBytes a, Typeable a) => FromJSONKey (UsingRawBytesHex a) where
+
+  fromJSONKey =
+    Aeson.FromJSONKeyTextParser $
+    either fail pure . deserialiseFromRawBytesBase16 . Text.encodeUtf8
+
+deserialiseFromRawBytesBase16 ::
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/SerialiseUsing.hs:89]: We now have two implementations of essentially the same thing `deserialiseFromRawBytesHex` and  `deserialiseFromRawBytesBase16`. Not to mention we Base16 decode the input `ByteString` twice in order to get an error message in `deserialiseFromRawBytesBase16`. The proper solution here is to update `deserialiseFromRawBytesHex` to return an `Either`. This refactor would be better as a separate commit.

File: cardano-api/src/Cardano/Api/Utils.hs [hunk 13]
```diff
@@ -42,3 +45,8 @@ runParsecParser parser input =
   case Parsec.parse (parser <* Parsec.eof) "" (Text.unpack input) of
     Right txin -> pure txin
     Left parseError -> fail $ formatParsecError parseError
+
+note :: MonadFail m => String -> Maybe a -> m a
```
Comment (Jimbo4350) [inline cardano-api/src/Cardano/Api/Utils.hs:49]: :+1:

File: cardano-api/src/Cardano/Api/Value.hs (hunk 4 above)
Comment (Jimbo4350) [reply cardano-api/src/Cardano/Api/Value.hs:367]: `note` needs to be removed from here

File: cardano-cli/src/Cardano/CLI/Helpers.hs [hunk 14]
```diff
@@ -126,3 +128,8 @@ hushM :: forall e m a. Monad m => Either e a -> (e -> m ()) -> m (Maybe a)
 hushM r f = case r of
   Right a -> return (Just a)
   Left e -> f e >> return Nothing
+
+note :: MonadFail m => String -> Maybe a -> m a
```
Comment (Jimbo4350) [inline cardano-cli/src/Cardano/CLI/Helpers.hs]: This needs to be deleted from here.

**Review by Jimbo4350 — APPROVED** [review]
Nice :+1:

**Comment by gitmachtl** [comment]
@JaredCorduan  @newhoggy maybe someone of you is having the time to review it also?

File: cardano-api/src/Cardano/Api/SerialiseUsing.hs (hunk 12 above)
Comment (cblp) [reply cardano-api/src/Cardano/Api/SerialiseUsing.hs:89]: Here it is #3304

**Comment by gitmachtl** [comment]
do we have a change to get this into the 1.31.0 release?

**Comment by gitmachtl** [comment]
@nc6 do we have a change to get this into the 1.31.0 release?

**Comment by cblp** [comment]
bors merge

**Comment by gitmachtl** [comment]
wohoo! :-)
